cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,382 @@
+            and self.measurement_key_map == other.measurement_key_map
+            and self.param_resolver == other.param_resolver
+            and self.repetitions == other.repetitions
+        )
+
+    # Methods for getting post-mapping properties of the contained circuit.
+
+    @property
+    def qubits(self) -> Tuple['cirq.Qid', ...]:
+        """Returns the qubits operated on by this object."""
+        ordered_qubits = ops.QubitOrder.DEFAULT.order_for(self.circuit.all_qubits())
+        return tuple(self._qubit_map.get(q, q) for q in ordered_qubits)
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return tuple(q.dimension for q in self.qubits)
+
+    def _measurement_keys_(self) -> AbstractSet[str]:
+        return {
+            self.measurement_key_map.get(key, key) for key in self.circuit.all_measurement_keys()
+        }
+
+    def _parameter_names_(self) -> AbstractSet[str]:
+        return {
+            name
+            for symbol in protocols.parameter_symbols(self.circuit)
+            for name in protocols.parameter_names(
+                protocols.resolve_parameters(symbol, self.param_resolver)
+            )
+        }
+
+    def _decompose_(self) -> 'cirq.OP_TREE':
+        result = self.circuit.unfreeze()
+        result = result.transform_qubits(lambda q: self.qubit_map[q])
+        if self.repetitions < 0:
+            result = result ** -1
+        result = protocols.with_measurement_key_mapping(result, self.measurement_key_map)
+        result = protocols.resolve_parameters(result, self.param_resolver)
+
+        return list(result.all_operations()) * abs(self.repetitions)
+
+    # Methods for string representation of the operation.
+
+    def __repr__(self):
+        result = f'cirq.CircuitOperation({self.circuit!r})'
+        base_op = self.base_operation()
+
+        def dict_repr(d: Dict) -> str:
+            pairs = [f'    {proper_repr(k)}: {proper_repr(v)},' for k, v in sorted(d.items())]
+            return '\n'.join(['{', *pairs, '}'])
+
+        if self.qubit_map != base_op.qubit_map:
+            result += f'.with_qubit_mapping({dict_repr(self.qubit_map)})'
+        if self.measurement_key_map != base_op.measurement_key_map:
+            result += f'.with_measurement_key_mapping({dict_repr(self.measurement_key_map)})'
+        if self.param_resolver != base_op.param_resolver:
+            result += f'.with_params({proper_repr(self.param_resolver)})'
+        if self.repetitions != 1:
+            result += f'.repeat({self.repetitions})'
+        return result
+
+    def __str__(self):
+        # TODO: support out-of-line subcircuit definition in string format.
+        header = self.circuit.serialization_key() + ':'
+        msg_lines = str(self.circuit).split('\n')
+        msg_width = max([len(header) - 4] + [len(line) for line in msg_lines])
+        circuit_msg = '\n'.join(
+            ['[ {line:<{width}} ]'.format(line=line, width=msg_width) for line in msg_lines]
```

Comment:
RE circuit diagrams: I intentionally avoided spending too much effort on constructing these since the intent is to replace them with out-of-line diagrams ASAP. Unless you see this as an urgent readability issue, I think we should postpone any diagram-related fixes until the out-of-line update rolls out.

Reply:
For reference, the need for out-of-line diagrams is captured in #3605.